homeassistant/components/ring/sensor.py

```diff
@@ -21,221 +30,209 @@
-        super()._handle_coordinator_update()
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        if self._latest_event is None:
-            return None
 
-        return self._latest_event["created_at"]
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        attrs = super().extra_state_attributes
+        self._device = cast(
+            _RingGenericT,
+            self._get_coordinator_data().get_device(self._device.device_api_id),
+        )
+        # History values can drop off the last 10 events so only update
+        # the value if it's not None
+        if native_value := self.entity_description.value_fn(self._device):
+            self._attr_native_value = native_value
+        if extra_attrs := self.entity_description.extra_state_attributes_fn(
+            self._device
+        ):
+            self._attr_extra_state_attributes = extra_attrs
+        super()._handle_coordinator_update()
 
-        if self._latest_event:
-            attrs["created_at"] = self._latest_event["created_at"]
-            attrs["answered"] = self._latest_event["answered"]
-            attrs["recording_status"] = self._latest_event["recording"]["status"]
-            attrs["category"] = self._latest_event["kind"]
 
-        return attrs
+def _get_last_event(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if not history_data:
+        return None
+    if kind is None:
+        return history_data[0]
+    for entry in history_data:
+        if entry["kind"] == kind.value:
+            return entry
+    return None
+
+
+def _get_last_event_attrs(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if last_event := _get_last_event(history_data, kind):
+        return {
+            "created_at": last_event.get("created_at"),
+            "answered": last_event.get("answered"),
+            "recording_status": last_event.get("recording", {}).get("status"),
+            "category": last_event.get("kind"),
+        }
+    return None
 
 
 @dataclass(frozen=True, kw_only=True)
-class RingSensorEntityDescription(SensorEntityDescription):
+class RingSensorEntityDescription(SensorEntityDescription, Generic[_RingGenericT]):
     """Describes Ring sensor entity."""
 
-    category: list[str]
-    cls: type[RingSensor]
+    value_fn: Callable[[_RingGenericT], StateType] = lambda _: True
+    exists_fn: Callable[[RingGeneric], bool] = lambda _: True
+    extra_state_attributes_fn: Callable[[_RingGenericT], dict[str, Any] | None] = (
+        lambda _: None
+    )
 
-    kind: str | None = None
 
-
-SENSOR_TYPES: tuple[RingSensorEntityDescription, ...] = (
+SENSOR_TYPES: tuple[
+    RingSensorEntityDescription[RingGeneric]
+    | RingSensorEntityDescription[RingDoorBell | RingChime]
+    | RingSensorEntityDescription[RingOther],
+    ...,
+] = (
```

Comment:
```suggestion
SENSOR_TYPES: tuple[RingSensorEntityDescription[Any, ...] = (
```
Quite a tricky situation here. This actually needs to be `Any, ...` for the `RingSensor(..., description)` call to be correct. As you noticed though, it causes the inference for `value_fn` to break. Not sure there is a good solution here.

It doesn't work correctly with the current annotations either. Just an example, for the `battery` description, `device` inside the `value_fn` lambda would be inferred as `RingDoorBell | RingChime` by both mypy and pyright. _I honestly don't know why._

Anyway, we somehow have to make it work now. I would suggest using `Any` but then for each `RingSensorEntityDescription` adding a subscript even if it should have been provided by the TypeVar default. That should do.

Reply:
Ah yes, so it did, that's strange.  Anyway I've added the subscripts, reverted to `Any` and put a comment to explain.